### PR TITLE
Add Spark backend for processing prio data

### PR DESCRIPTION
**Pull request by acmiyaguchi**
Now that multiprocessing works in the Python bindings, Spark is a viable option for parallelizing jobs on a cluster. The Spark jobs will reuse the same interface as the existing command-line interface, and should be transparent since the application is deployed as a container. 

This requires both #93 and #94. The latter is important for the tests to pass, since it updates the test fixtures for the new secret-shared seed mechanism.

### CODE REVIEW

File: prio_processor/spark/commands.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+            lambda pdf: udf.encode(
+                batch_id, n_data, public_key_hex_internal, public_key_hex_external, pdf
+            ),
+            schema="a: binary, b: binary",
+        )
+        .withColumn("id", F.udf(lambda: str(uuid4()), returnType="string")())
+    )
+    shares.cache()
+
+    # should this guarantee valid ranges for each partition?
```
Comment (wlach) [inline prio_processor/spark/commands.py:136]: This comment is a bit concerning. I assume the answer is yes? 😃

File: prio_processor/spark/commands.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    # is the same here.
+    row = shares.first()
+    num_partitions = math.ceil(
+        (len(b64encode(row.shares.a)) + len(str(uuid4())))
+        * n_rows
+        * scale
+        * 1.0
+        / 10 ** 6
+        / partition_size_mb
+    )
```
Comment (wlach) [inline prio_processor/spark/commands.py:93]: I think splitting this up into two steps makes it easier to understand:

```suggestion
    data_size = math.ceil(
        (len(b64encode(row.shares.a)) + len(str(uuid4())))
        * n_rows
        * scale
        * 1.0
        / 10 ** 6
    )
    num_partitions = data_size / partition_size_mb
```

File: prio_processor/spark/commands.py (hunk 1 above)
Comment (acmiyaguchi) [reply prio_processor/spark/commands.py:136]: Yeah, but it was less clear before it was using the range partitioning.

File: prio_processor/spark/udf.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+
+def encode(
+    batch_id: bytes,
+    n_data: int,
+    public_key_hex_internal: bytes,
+    public_key_hex_external: bytes,
+    payload: pd.Series,
+) -> pd.DataFrame:
+    with PrioContext():
```
Comment (wlach) [inline prio_processor/spark/udf.py:19]: How often is this called? Could it add significant overhead in some cases?

File: prio_processor/spark/udf.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    shared_secret: bytes,
+    public_key_hex_internal: bytes,
+    public_key_hex_external: bytes,
+    shares: pd.Series,
+    internal: pd.Series,
+    external: pd.Series,
+) -> pd.Series:
+    libprio.Prio_init()
+
+    def _process(share, internal, external):
```
Comment (wlach) [inline prio_processor/spark/udf.py:88]: Quite a bit of duplication between this and `_process` in verify1. Could you at least factor out the first part of this function (up to where you create the verifier variable) into a common function?

File: prio_processor/spark/udf.py (hunk 3 above)
Comment (acmiyaguchi) [reply prio_processor/spark/udf.py:19]: This is called once per arrow batch, which by default something like 2-4 GB. It's necessary overhead, but I don't believe it does too much other than initializing the NSS context and setting up the random number generator.

**Pull request merged**